src/ffi/buffer.rs

```diff
@@ -0,0 +1,242 @@
+//! AssemblyScript dynamically sized buffer implementation.
+
+use std::{
+    alloc::{self, Layout, LayoutErr},
+    fmt::{self, Debug, Formatter},
+    marker::PhantomData,
+    mem,
+    ops::Deref,
+    ptr, slice,
+};
+
+/// Internal representation of an AssemblyScript buffer.
+///
+/// `Inner` is declared as a generic struct in order to take advantage of the
+/// partial dynamically sized type (DST) support. For more information see:
+/// <https://doc.rust-lang.org/nomicon/exotic-sizes.html#dynamically-sized-types-dsts>
+#[repr(C)]
+struct Inner<T: ?Sized> {
+    len: usize,
+    buf: T,
+}
+
+/// A borrowed AssemblyScript dynamically sized buffer with elements of type `T`
+/// and aligned to `Alignment`.
+#[repr(transparent)]
+pub struct AscBuf<T, Alignment = T> {
+    _type: PhantomData<T>,
```

Comment:
Judging by the variable name this isn't used. What do you expect it to be used for?

Reply:
It is needed for a couple of reasons:
- Rust requires all type parameters to be used. So we just use a `PhantomData` which is a 0-size marker type.
- `PhatomData` acts like it "owns a `T`". This is important for auto-traits (`Sync`, `Send`, `Unpin`, etc.).